pkg/queryparser/sql/sql.go

```diff
@@ -0,0 +1,654 @@
+		},
+		"NCONTAINS": {
+			usedBy:        queryparser.NewSet[string]().Add(queryparser.RootFunc, "AND", "OR"),
+			Verifications: []verificationHandler{withKeyArg(), withNoValues()},
+			handle:        Wrap(sp.queryNotContains),
+		},
+		"OVERLAPS": {
+			usedBy:        queryparser.NewSet[string]().Add(queryparser.RootFunc, "AND", "OR"),
+			Verifications: []verificationHandler{withKeyArg(), withNoValues()},
+			handle:        Wrap(sp.queryOverlaps),
+		},
+		"NOVERLAPS": {
+			usedBy:        queryparser.NewSet[string]().Add(queryparser.RootFunc, "AND", "OR"),
+			Verifications: []verificationHandler{withKeyArg(), withNoValues()},
+			handle:        Wrap(sp.queryNotOverlaps),
+		},
+		"ANY": {
+			usedBy:        queryparser.NewSet[string]().Add(queryparser.RootFunc, "AND", "OR"),
+			Verifications: []verificationHandler{withKeyArg(), withNoValues()},
+			handle:        Wrap(sp.queryAny),
+		},
+		"NANY": {
+			usedBy:        queryparser.NewSet[string]().Add(queryparser.RootFunc, "AND", "OR"),
+			Verifications: []verificationHandler{withKeyArg(), withNoValues()},
+			handle:        Wrap(sp.queryNotAny),
+		},
+		"ALL": {
+			usedBy:        queryparser.NewSet[string]().Add(queryparser.RootFunc, "AND", "OR"),
+			Verifications: []verificationHandler{withKeyArg(), withNoValues()},
+			handle:        Wrap(sp.queryAll),
+		},
+		"NALL": {
+			usedBy:        queryparser.NewSet[string]().Add(queryparser.RootFunc, "AND", "OR"),
+			Verifications: []verificationHandler{withKeyArg(), withNoValues()},
+			handle:        Wrap(sp.queryNotAll),
+		},
+		"CAST": {
+			usedBy: queryparser.NewSet[string]().Add("EQ", "NEQ", "LT", "LTE", "GT", "GTE", "IN", "NOTIN", "LIKE",
+				"NLIKE", "ANY", "NANY", "ALL", "NALL", "OVERLAPS", "NOVERLAPS", "CONTAINS", "NCONTAINS", "ISNULL", "ISNOTNULL"),
+			Verifications: []verificationHandler{withKeyOrValueArg()},
+			handle:        sp.queryCast,
+		},
+		"K": {
+			usedBy: queryparser.NewSet[string]().Add("EQ", "NEQ", "LT", "LTE", "GT", "GTE", "IN", "NOTIN", "LIKE",
+				"NLIKE", "ANY", "NANY", "ALL", "NALL", "OVERLAPS", "NOVERLAPS", "CONTAINS", "NCONTAINS", "ISNULL", "ISNOTNULL", "CAST"),
+			handle: Wrap(sp.queryKey),
+		},
+		"V": {
+			usedBy: queryparser.NewSet[string]().Add("EQ", "NEQ", "LT", "LTE", "GT", "GTE", "IN", "NOTIN", "LIKE",
+				"NLIKE", "ANY", "NANY", "ALL", "NALL", "OVERLAPS", "NOVERLAPS", "CONTAINS", "NCONTAINS", "CAST"),
+			handle: sp.queryValue,
+		},
+	}
+
+	for _, option := range options {
+		if err := option(sp); err != nil {
+			return nil, err
+		}
+	}
+	return sp, nil
+}
+
+type parser struct {
+	sqlparser *SQLParser
+}
+
+// Parse constructs a SQL query based on the provided input.
+// This method tokenizes the input, verifies the structure of the tokens,
+// and executes the corresponding SQL functions to generate the final query
+// string along with its parameters.
+func (sp *SQLParser) Parse(ctx context.Context, input any, params ...string) (string, []any, error) {
+	if input == "" {
```

Comment:
nit: If input can be any then we should use type assertion to ensure `input.(string)` before we evaluate it as a string.

Reply:
Hey,

I will fix it and compare it to nil.
The input can be anything, for the field selector is a type selector that will be processed by the tokenizer.
In this case, the field selector implements a tokenizer and doesn't use the default.
Therefore, the assertion can be found here: https://github.com/flightctl/flightctl/blob/9e7765f7d44602c5757d1414be45ebc3094a858e/internal/store/selector/field.go#L80

In case when the default tokenizer is being used: https://github.com/flightctl/flightctl/blob/9e7765f7d44602c5757d1414be45ebc3094a858e/pkg/queryparser/queryparser.go#L299

Initially, it was indeed string until more types of inputs were introduced.